### DISCUSSION TITLE
Are there any working examples/repos that follow the guidelines provided here?

### DISCUSSION THREAD

**Discussion by nareshbhatia** [graphql-rules/graphql-rules#123] [Q&A] [answered]
I am sure many people will find it useful. For example, I am trying to figure out how to name the top level queries for my types. I see implicitly in the examples here that tip level list queries are pascalCase plurals of the type name, e.g. `articles`. Unfortunately, there are no examples of getting one object by its id. I assume such a query should be called `article` (singular).

Some examples on the net append `_Query` to get the query name, e.g. `Article_Query`. But that doesn't work well with `graphql-codegen` because it appends the word `Query` to the generated types, e.g. `Article_QueryQuery`.

Would be really nice to see how people make all these tools work together.

**Comment by RIP21** [comment]
That would be a lot of work to do a whole course TBH to explain these all in the real world or at least good examples :)
I think I could do such a course but that's hours of hours of work :)

But the best approach is to call it like `query Article_ArticlePage`
Basically, after the component/hook/fetcher name that hosts/executes it. So, let's say if it's a React Hook then `Article_useArticle` or so.

Unsure how Relay Compiler names queries (I recommend people to use Relay yet never used it in production myself lol), but probably similarly :)

**Reply by nareshbhatia** [reply]
@RIP21, your response gave me an aha moment. What I was referring to as the keyword `query` is used in multiple places:
1. When defining a `type query` in the schema
2. When defining a `query` on a page

This is the specific example I am working on (it is an attempt to create working code for [Laurin Quast's example](https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen)):

```graphql
# example-schema.graphql

type User {
  id: ID!
  fullName: String!
  avatarUrl: String!
  homeTown: String!
  registeredAt: String!
  friends: [User!]!
}

type Query {
  user(id: ID!): User
}
```

```tsx
// page.tsx

import { graphql } from '@/generated/gql';
import { useQuery } from '@apollo/client';

const userId = 'john.smith';

const User_UserProfileRouteQuery = graphql(/* GraphQL */ `
  query User_UserProfile($id: ID!) {
    user(id: $id) {
      id
      fullName
    }
  }
`);

export default function UserProfile() {
  const { data } = useQuery(User_UserProfileRouteQuery, {
    variables: {
      id: userId,
    },
  });
  const user = data?.user;
  if (!user) return undefined;

  return (
    <div className="mx-auto max-w-3xl p-4">
      {user.id} {user.fullName}
    </div>
  );
}
```

As you can see the query has different names in different places:
1. In `example-schema.graphql`, it is called `user`
2. In `page.tsx`, the typescript variable is called `User_UserProfileRouteQuery` whereas the query inside is called `User_UserProfile`.

I am sure these all have different semantic meanings and do different things, but I don't understand them yet. Would be good to understand. In any case, I would like to adopt good naming conventions for each one of these.

BTW, graphql-codegen generates the following types from `User_UserProfile`, so that name seems to be important for it:
1. `User_UserProfileQuery`
2. `User_UserProfileQueryVariables`
4. `User_UserProfileDocument`

**Reply by nareshbhatia** [reply]
Just found [this example](https://github.com/dotansimha/graphql-code-generator/blob/master/examples/react/apollo-client/src/App.tsx) in graphql-codegen repo. It names like this:
1. Typescript variable: `allFilmsWithVariablesQueryDocument`
2. Query inside it: `allFilmsWithVariablesQuery`
3. The actual query in the schema: `allFilms`

**Comment by nareshbhatia** [comment, marked as answer]
I finally created two examples to show how [graphql-code-generator](https://github.com/dotansimha/graphql-code-generator) & [graphql-eslint](https://github.com/dimaMachina/graphql-eslint) can be used together to follow the guidelines in this repository. Note that `graphql-code-generator` is very intricate in the way it generates code from the names specified in your code (see example [here](https://github.com/nareshbhatia/graphql-fragments-example-1/blob/main/src/app/page.tsx#L10-L24)).

Here are my two example repositories:

1. https://github.com/nareshbhatia/graphql-fragments-example-1
2. https://github.com/nareshbhatia/graphql-fragments-example-2
